### PR TITLE
Add ctor/dtor/copy/assign to std::vector bindings

### PR DESCRIPTION
**Pull request by omerfirmak**

### CODE REVIEW

**Comment by omerfirmak** [comment]
```
submodules/serialization/source/agora/serialization/Serializer.d(698,17): Error: none of the overloads of __ctor are callable using a const object, candidates are:
source/scpd/Cpp.d(62,9):        scpd.types.Stellar_SCP.SCPBallot.CPPBindingMixin!(SCPBallot, true, true).this(CppCtor use)
source/scpd/Cpp.d(57,13):        scpd.types.Stellar_SCP.SCPBallot.CPPBindingMixin!(SCPBallot, true, true).this(uint, xvector!(ubyte, 4294967292u))
source/scpd/Cpp.d(73,9):        scpd.types.Stellar_SCP.SCPBallot.CPPBindingMixin!(SCPBallot, true, true).this(return ref scope inout(SCPBallot) rhs)

```

Shouldnt this work with `const` objects?

https://github.com/bosagora/agora/pull/2081/files#diff-2fff5f741967021f44722f88e37effa07123b287d13e843ff246b43d91a7d8d4R57-R60

**Comment by omerfirmak** [comment]
This mostly works apart from a random crash that happens during GC finalizing a `vector`

**Comment by omerfirmak** [comment]
> This mostly works apart from a random crash that happens during GC finalizing a `vector`

I can't reproduce this after a rebase but I don't see where we would GC allocate a `SCPEnvelope` anyways. Marking this ready for review for the time being.

File: source/scpd/Cpp.d [hunk 1]
```diff
@@ -37,44 +37,44 @@ import vibe.data.json;
...
-    void defaultCtorCPPObject(T) (T* ptr);
-    void dtorCPPObject(T) (T* ptr);
-    void opAssignCPPObject(T) (T* lhs, T* rhs);
-    void copyCtorCPPObject(T) (T* ptr, inout(T)* rhs);
-    void copyCtorCPPObject(T) (immutable(T)* ptr, inout(T)* rhs);
+@trusted nothrow @nogc pure:
+    void defaultCtorCPPObject(T) (inout(T)* ptr);
+    void dtorCPPObject(T) (inout(T)* ptr);
+    void opAssignCPPObject(T) (inout(T)* lhs, inout(T)* rhs);
+    void copyCtorCPPObject(T) (inout(T)* ptr, inout(T)* rhs);
```
Comment (Geod24) [inline source/scpd/Cpp.d:44]: Not sure why you would need `inout` for single-argument templated free functions, but I'm sure there's an horror story behind this :)

File: source/scpd/Cpp.d [hunk 2]
```diff
@@ -37,44 +37,44 @@ import vibe.data.json;
...
+public mixin template CPPBindingMixin (T, bool Copyable = true, bool DefaultConstructable = false)
 {
     extern(D)
     {
-        static if (!DefaultConstructable) @disable this();
-        this (CppCtor use) @trusted nothrow @nogc
+        import std.traits : Fields, FieldNameTuple;
+
+        static if (!DefaultConstructable) 
+            @disable this();
```
Comment (Geod24) [inline source/scpd/Cpp.d]:
```suggestion
            @disable this ();
```
👼

File: source/scpd/Cpp.d [hunk 3]
```diff
@@ -37,44 +37,44 @@ import vibe.data.json;
...
-            copyCtorCPPObject!T(&this, &rhs);
-        }
-
-        static if (Copyable)
-        this (ref return scope inout T rhs) immutable @trusted nothrow @nogc
+        this (ref return scope inout T rhs) inout @trusted nothrow @nogc pure
         {
+            // Without this, compiler thinks we are not initializing anything
+            static foreach (field; FieldNameTuple!T)
+                __traits(getMember, this, field) = __traits(getMember, this, field).init;
```
Comment (Geod24) [inline source/scpd/Cpp.d]: Would it be possible to nest it in a dead condition just so that we don't run code that's not needed ?

**Review by Geod24 — CHANGES_REQUESTED** [review]

File: source/scpd/Cpp.d (hunk 3 above)
Comment (omerfirmak) [reply source/scpd/Cpp.d]: Yeah that worked.

**Comment by omerfirmak** [comment]
Rebased

**Review by Geod24 — APPROVED** [review]
Noice

**Comment by Geod24** [comment]
Reverted in #2143

**Pull request merged**